### PR TITLE
Add implementation of VK_KHR_synchronization2

### PR DESCRIPTION
**Pull request by jeremyg-lunarg**
Add a 'reference' implementation of this extension, which should work on Windows, Linux, MacOS and Android. This layer makes use of infrastructure from the Vulkan-ValidationLayer. To reduce overall repo dependency complexity, the relevant code is pulled in directly under utils/* and tests/*.

### CODE REVIEW

File: build-android/known_good.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      "name" : "glslang",
+      "url" : "https://github.com/KhronosGroup/glslang.git",
+      "sub_dir" : "shaderc/third_party/glslang",
+      "commit" : "dd69df7f3dac26362e10b0f38efb9e47990f7537"
+    },
+    {
+      "name" : "Vulkan-Headers",
+      "url" : "https://github.com/KhronosGroup/Vulkan-Headers.git",
+      "sub_dir" : "Vulkan-Headers",
+      "commit" : "v1.2.162"
```
Comment (sfricke-samsung) [inline build-android/known_good.json:19]: 170

File: utils/vk_layer_config.cpp [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+        separator = true;
+    }
+    if (vk_flags & VK_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT) {
+        if (separator) strcat(msg_flags, ",");
+        strcat(msg_flags, "PERF");
+    }
+}
+
+// This catches before dlopen fails if the default Android-26 layers are being used and attempted to be ran on Android 25 or below
+#if defined(__ANDROID__)
```
Comment (sfricke-samsung) [inline utils/vk_layer_config.cpp:420]: this chunk got added back in I see, need to remove

File: utils/vk_layer_config.cpp [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    }
+    return flags;
+}
+
+VK_LAYER_EXPORT void setLayerOption(const char *option, const char *value) { layer_config.SetOption(option, value); }
+
+// Constructor for ConfigFile. Initialize layers to log error messages to stdout by default. If a vk_layer_settings file is present,
+// its settings will override the defaults.
+ConfigFile::ConfigFile() : file_is_parsed_(false) {
+    value_map_["khronos_validation.report_flags"] = "error";
```
Comment (sfricke-samsung) [inline utils/vk_layer_config.cpp:180]: these seem not needed

File: BUILD.md [hunk 4]
```diff
@@ -36,6 +36,47 @@ the validation layers. You must also take note of the headers' install
...
 
+#### Google Test
+
+The extension layer tests depend on the
+[Google Test](https://github.com/google/googletest)
+framework and do not build unless this framework is downloaded into the
+repository's `external` directory.
+
+To obtain the framework, change your current directory to the top of your
+Vulkan-ValidationLayers repository and run:
```
Comment (sfricke-samsung) [inline BUILD.md:47]:
```suggestion
Vulkan-ExtensionLayer repository and run:
```

File: tests/extension_layer_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+            } else {
+                __android_log_print(ANDROID_LOG_INFO, appTag, "No Intent args detected");
+            }
+
+            int argc = 2;
+            char *argv[] = {(char *)"foo", (char *)filter.c_str()};
+            __android_log_print(ANDROID_LOG_DEBUG, appTag, "filter = %s", argv[1]);
+
+            // Route output to files until we can override the gtest output
+            freopen("/sdcard/Android/data/com.example.VulkanLayerValidationTests/files/out.txt", "w", stdout);
```
Comment (sfricke-samsung) [inline tests/extension_layer_tests.cpp:486]: fiix name

File: tests/vkrenderframework.cpp [hunk 6]
```diff
@@ -0,0 +1,2186 @@
...
+        DumpFailureMsgs();
+        for (const auto &desired_msg : desired_message_strings_) {
+            ADD_FAILURE() << "Did not receive expected error '" << desired_msg << "'";
+        }
+    } else if (GetOtherFailureMsgs().size() > 0) {
+        // Fail test case for any unexpected errors
+#if defined(ANDROID)
+        // This will get unexpected errors into the adb log
+        for (auto msg : other_messages_) {
+            __android_log_print(ANDROID_LOG_INFO, "VulkanLayerValidationTests", "[ UNEXPECTED_ERR ] '%s'", msg.c_str());
```
Comment (sfricke-samsung) [inline tests/vkrenderframework.cpp:193]: rename `VulkanLayerValidationTests`

if you search for `VulkanLayerValidationTests` you will find more

File: build-android/jni/Android.mk [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+include $(CLEAR_VARS)
+LOCAL_MODULE := VkLayer_khronos_synchronization
+LOCAL_SRC_FILES += $(SRC_DIR)/layers/synchronization2.cpp
+LOCAL_C_INCLUDES += $(VULKAN_INCLUDE) \
+                    $(LOCAL_PATH)/$(SRC_DIR)/utils \
+                    $(LOCAL_PATH)/$(SRC_DIR)/utils/generated \
+                    $(LOCAL_PATH)/$(THIRD_PARTY)/shaderc/third_party/spirv-tools/external/spirv-headers/include
+LOCAL_STATIC_LIBRARIES += extlayer_utils glslang SPIRV-Tools SPIRV-Tools-opt
+LOCAL_CPPFLAGS += -std=c++11 -Wall -Werror -Wno-unused-function -Wno-unused-const-variable -fexceptions
```
Comment (sfricke-samsung) [inline build-android/jni/Android.mk:47]: I needed `-Wno-cast-calling-convention` or else the 32-bit build got mad about `ADD_HOOK`

File: build-android/cmake/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
+Build Extension Layers with Android CMake Plugin
```
Comment (sfricke-samsung) [inline build-android/cmake/README.md:1]: I don't recall what anything in `build-android/cmake` is for

Are you using this? Main reason is it has a lot of `validation layer` references and not sure if this is even needed

**Review by sfricke-samsung — COMMENTED** [review]
Overall the sync2 layer logic seems to be good to go and all the previous internal issues are addressed

The only thing is some nit stuff related to building on Android and usage of "validation" still around... I decided to not to get too nit picky as realize there are still a few other plates spinning with sync2

**Comment by jeremyg-lunarg** [comment]
Thanks for reviewing! I think I got all of these addressed.

I'm kicking off internal CI on this PR again to make sure I didn't break windows. It probably doesn't make sense for you to look again until this finishes (I'll post again when it does).

 I'm going to go back and double check on github that I got all your comments.

File: utils/vk_layer_config.cpp (hunk 3 above)
Comment (jeremyg-lunarg) [reply utils/vk_layer_config.cpp]: I only partially got this one.

File: build-android/cmake/README.md (hunk 8 above)
Comment (jeremyg-lunarg) [reply build-android/cmake/README.md:1]: not needed. I'm not sure why VVL has it.

**Comment by jeremyg-lunarg** [comment]
updated, double checked and passing Win32 on internal CI

**Pull request merged**